File: BKabanApi/BKabanApi/DB/ColumnRepository.cs
using System.Data;
using System.Diagnostics;
using System.Linq;
using Dapper;
using Microsoft.Data.SqlClient;

namespace BKabanApi.Models.DB
{
    public interface IColumnRepository
    {
        int? CreateColumn(int userId, ColumnModelBoardLink column);
        bool UpdateColumn(int userId, ColumnModel column);
        bool UpdateColumnPosition(int userId, ColumnModelWithPosition column);
        bool DeleteColumn(int userId, int id);
    }

    public class ColumnRepository : IColumnRepository
    {
        private readonly string _connectionString;

        public ColumnRepository(string connectionString)
        {
            _connectionString = connectionString;
        }

        public static bool IsAllowed(IDbConnection db, int userId, int columnId)
        {
            return db.Query(@"SELECT C.Id
                FROM UserTable AS U
                JOIN BoardTable AS B ON B.UserId=U.Id
                JOIN ColumnTable AS C ON C.BoardId=B.Id
                WHERE U.Id = @userId AND C.Id=@columnId", new { userId, columnId }).Any();
        }

        private static int GetColumnsCount(IDbConnection db, int boardId)
        {
            return db.Query<int>(
                @"SELECT COUNT(C.Id)
                FROM BoardTable AS B
                JOIN ColumnTable AS C ON B.Id = C.BoardId
                WHERE B.Id = @boardId",
                new { boardId }).FirstOrDefault();
        }

        public int? CreateColumn(int userId, ColumnModelBoardLink column)
        {
            using IDbConnection db = new SqlConnection(_connectionString);
            Debug.Assert(column.BoardId != null, "column.BoardId != null");
            if (!BoardRepository.IsAllowed(db, userId, (int)column.BoardId))
            {
                return null;
            }

            return db.Query<int?>(
                @"INSERT INTO ColumnTable(Name, BoardId, Position) 
                OUTPUT INSERTED.Id VALUES
                (@columnName, @boardId, @position);",
                new { boardId = column.BoardId, columnName = column.Name, position = GetColumnsCount(db, (int)column.BoardId) }).FirstOrDefault();
        }

        public bool UpdateColumn(int userId, ColumnModel column)
        {
            using IDbConnection db = new SqlConnection(_connectionString);
            Debug.Assert(column.Id != null, "column.Id != null");
            if (!IsAllowed(db, userId, (int)column.Id))
            {
                return false;
            }

            return db.Query(
                @"UPDATE ColumnTable
                SET Name = @newName
                OUTPUT DELETED.Id
                WHERE Id = @id", new { newName = column.Name, id = column.Id }).Any();
        }

        public bool UpdateColumnPosition(int userId, ColumnModelWithPosition column)
        {
            using IDbConnection db = new SqlConnection(_connectionString);
            Debug.Assert(column.Id != null, "column.Id != null");
            if (!IsAllowed(db, userId, (int)column.Id))
            {
                return false;
            }

            int? boardId = db.Query<int?>(
                @"SELECT B.Id 
	            FROM ColumnTable AS C
	            JOIN BoardTable AS B ON B.Id = C.BoardId 
	            WHERE C.Id = @columnId;",
                new { columnId = column.Id }).FirstOrDefault();

            if (boardId == null)
            {
                return false;
            }

            int maxColumnPosition = GetColumnsCount(db, (int) boardId) - 1;

            if (column.Position < 0 || column.Position > maxColumnPosition)
            {
                return false;
            }

            return db.Query(
                @"DECLARE @OldColumnPosition INT;
                SET @OldColumnPosition = 
	                (SELECT Position 
	                FROM ColumnTable 
	                WHERE Id = @columnId);

                UPDATE ColumnTable
                SET Position = C.Position - 1
                FROM ColumnTable AS C
                JOIN BoardTable AS B ON B.Id = C.BoardId
                WHERE B.Id = @boardId
                AND C.Position > @OldColumnPosition;

                UPDATE ColumnTable
                SET Position = C.Position + 1
                FROM ColumnTable AS C
                JOIN BoardTable AS B ON B.Id = C.BoardId
                WHERE B.Id = @boardId
                AND C.Position >= @newPosition;

                UPDATE ColumnTable
                SET Position = @newPosition
                OUTPUT DELETED.Id
                WHERE Id = @columnId;",
                new { boardId, columnId = column.Id, newPosition = column.Position}).Any();
        }

        public bool DeleteColumn(int userId, int id)
        {
            using IDbConnection db = new SqlConnection(_connectionString);
            if (!IsAllowed(db, userId, id))
            {
                return false;
            }

            return db.Query(
                @"DECLARE @BoardId INT, @ColumnPosition INT;
                SET @BoardId =
	                (SELECT B.Id 
	                FROM ColumnTable AS C
	                JOIN BoardTable AS B ON B.Id = C.BoardId 
	                WHERE C.Id = @columnId);
                SET @ColumnPosition = 
	                (SELECT Position 
	                FROM ColumnTable 
	                WHERE Id = @columnId);

                UPDATE ColumnTable
                SET Position = C.Position - 1
                FROM ColumnTable AS C
                JOIN BoardTable AS B ON B.Id = C.BoardId
                WHERE B.Id = @BoardId
                AND C.Position > @ColumnPosition

                DELETE ColumnTable
                OUTPUT DELETED.Id
                WHERE Id = @columnId",
                new { columnId = id }).Any();
        }
    }
}

File: BKabanApiJava/src/main/java/edu/bkaban/repositories/ColumnRepository.java
package edu.bkaban.repositories;

import edu.bkaban.models.column.ColumnModel;
import edu.bkaban.models.column.ColumnModelBoardLink;
import edu.bkaban.models.column.ColumnModelWithPosition;
import edu.bkaban.services.DbService;
import org.springframework.stereotype.Service;

import java.sql.Connection;
import java.sql.SQLException;

@Service
public class ColumnRepository {
    private static final String isAllowedSql =
            "SELECT C.Id\n" +
                    "FROM UserTable AS U\n" +
                    "JOIN BoardTable AS B ON B.UserId=U.Id\n" +
                    "JOIN ColumnTable AS C ON C.BoardId=B.Id\n" +
                    "WHERE U.Id = ? AND C.Id = ?";

    public static boolean isAllowed(Connection conn, int userId, int columnId) throws SQLException {
        try (var stmt = conn.prepareStatement(isAllowedSql)) {
            stmt.setInt(1, userId);
            stmt.setInt(2, columnId);

            return stmt.executeQuery().next();
        }
    }

    private static final String getColumnsCountSql =
            "SELECT COUNT(C.Id)\n" +
                    "FROM BoardTable AS B\n" +
                    "JOIN ColumnTable AS C ON B.Id = C.BoardId\n" +
                    "WHERE B.Id = ?";

    private static int getColumnsCount(Connection conn, int boardId) throws SQLException {
        try (var stmt = conn.prepareStatement(getColumnsCountSql)) {
            stmt.setInt(1, boardId);

            var rset = stmt.executeQuery();
            rset.next();
            return rset.getInt(1);
        }
    }

    private final String createColumnSql =
            "INSERT INTO ColumnTable(Name, BoardId, Position) \n" +
                    "OUTPUT INSERTED.Id VALUES\n" +
                    "(?, ?, ?)";

    public Integer createColumn(int userId, ColumnModelBoardLink column) throws SQLException {
        try (var conn = DbService.getConnection(); var stmt = conn.prepareStatement(createColumnSql)) {
            if (!BoardRepository.isAllowed(conn, userId, column.getBoardId())) {
                return null;
            }

            stmt.setString(1, column.getName());
            stmt.setInt(2, column.getBoardId());
            stmt.setInt(3, getColumnsCount(conn, column.getBoardId()));

            var rset = stmt.executeQuery();
            if (!rset.next()) {
                return null;
            }
            return rset.getInt(1);
        }
    }

    private final String updateColumnSql =
            "UPDATE ColumnTable\n" +
                    "SET Name = ?\n" +
                    "OUTPUT DELETED.Id\n" +
                    "WHERE Id = ?";

    public boolean updateColumn(int userId, ColumnModel column) throws SQLException {
        try (var conn = DbService.getConnection(); var stmt = conn.prepareStatement(updateColumnSql)) {
            if (!isAllowed(conn, userId, column.getId())) {
                return false;
            }

            stmt.setString(1, column.getName());
            stmt.setInt(2, column.getId());

            return stmt.executeQuery().next();
        }
    }

    private final String getParentBoardSql =
            "SELECT B.Id \n" +
                    "FROM ColumnTable AS C\n" +
                    "JOIN BoardTable AS B ON B.Id = C.BoardId\n" +
                    "WHERE C.Id = ?;";

    private final String updateColumnPositionSql =
            "DECLARE @OldColumnPosition INT;\n" +
                    "SET @OldColumnPosition = \n" +
                    "(SELECT Position \n" +
                    "FROM ColumnTable \n" +
                    "WHERE Id = ?);\n" +
                    "UPDATE ColumnTable\n" +
                    "SET Position = C.Position - 1\n" +
                    "FROM ColumnTable AS C\n" +
                    "JOIN BoardTable AS B ON B.Id = C.BoardId\n" +
                    "WHERE B.Id = ?\n" +
                    "AND C.Position > @OldColumnPosition;\n" +
                    "UPDATE ColumnTable\n" +
                    "SET Position = C.Position + 1\n" +
                    "FROM ColumnTable AS C\n" +
                    "JOIN BoardTable AS B ON B.Id = C.BoardId\n" +
                    "WHERE B.Id = ?\n" +
                    "AND C.Position >= ?;\n" +
                    "UPDATE ColumnTable\n" +
                    "SET Position = ?\n" +
                    "OUTPUT DELETED.Id\n" +
                    "WHERE Id = ?;";

    public boolean updateColumnPosition(int userId, ColumnModelWithPosition column) throws SQLException {
        try (var conn = DbService.getConnection(); var stmt = conn.prepareStatement(updateColumnPositionSql); var boardIdStmt = conn.prepareStatement(getParentBoardSql)) {
            if (!isAllowed(conn, userId, column.getId())) {
                return false;
            }

            boardIdStmt.setInt(1, column.getId());
            var boardIdRset = boardIdStmt.executeQuery();

            if (!boardIdRset.next()) {
                return false;
            }

            int boardId = boardIdRset.getInt(1);

            int maxColumnPosition = getColumnsCount(conn, boardId) - 1;

            if (column.getPosition() < 0 || column.getPosition() > maxColumnPosition) {
                return false;
            }

            stmt.setInt(1, column.getId());
            stmt.setInt(2, boardId);
            stmt.setInt(3, boardId);
            stmt.setInt(4, column.getPosition());
            stmt.setInt(5, column.getPosition());
            stmt.setInt(6, column.getId());


            return stmt.executeQuery().next();
        }
    }

    private final String deleteColumnSql =
            "DECLARE @BoardId INT, @ColumnPosition INT;\n" +
                    "SET @BoardId =\n" +
                    "(SELECT B.Id \n" +
                    "FROM ColumnTable AS C\n" +
                    "JOIN BoardTable AS B ON B.Id = C.BoardId \n" +
                    "WHERE C.Id = ?);\n" +
                    "SET @ColumnPosition = \n" +
                    "(SELECT Position \n" +
                    "FROM ColumnTable \n" +
                    "WHERE Id = ?);\n" +
                    "UPDATE ColumnTable\n" +
                    "SET Position = C.Position - 1\n" +
                    "FROM ColumnTable AS C\n" +
                    "JOIN BoardTable AS B ON B.Id = C.BoardId\n" +
                    "WHERE B.Id = @BoardId\n" +
                    "AND C.Position > @ColumnPosition\n" +
                    "DELETE ColumnTable\n" +
                    "OUTPUT DELETED.Id\n" +
                    "WHERE Id = ?";

    public boolean deleteColumn(int userId, int id) throws SQLException {
        try (var conn = DbService.getConnection(); var stmt = conn.prepareStatement(deleteColumnSql)) {
            if (!isAllowed(conn, userId, id)) {
                return false;
            }

            stmt.setInt(1, id);
            stmt.setInt(2, id);
            stmt.setInt(3, id);

            return stmt.executeQuery().next();
        }
    }

}
